File: iOS/Objective-C/Lab2/iOS-Lab-2/ViewController.m
//
//  ViewController.m
//  
//
//  Created by Maha Basheikh on 2023-09-20.
//

#import "ViewController.h"
#import "CalBrain.h" // import the calBrian implementation

@interface ViewController ()
// create a display object
@property(weak,nonatomic)IBOutlet UILabel *myDisplay;

// create a boolean value
@property bool isNewNumber;

// create CalBrain object myModel
@property (nonatomic)CalBrain * myModel;
@end

@implementation ViewController

//lazy loading (only loads when the getter for myModel is used)
-(CalBrain*)myModel{

if (_myModel == nil){
    // allocate memory for mymodel obj
    _myModel =  [[CalBrain alloc]init];

}
    return _myModel;
}
// create a button which shows the button value in the display when pressed
- (IBAction)DigitPressed:(UIButton *)sender {
    // Display if new number
    if(self.isNewNumber){
        // Display the entered number
        self.myDisplay.text = sender.titleLabel.text;
        // Set the isNumber to No
        self.isNewNumber = NO;
    }else{
        // if not new number assing string value to what is shown in the display
        NSString *valueOnDisplay = self.myDisplay.text;
        
        // assing string to the new digit entered
        NSString *newDigit = sender.titleLabel.text;
        
        // save both nubmeres in string format (concatenate)
        NSString *newNumber = [NSString stringWithFormat:@"%@%@",valueOnDisplay,newDigit];
        
        // display the new concatenated value
        self.myDisplay.text = newNumber;
    }
}

// Function to calculate the result based on the operator
- (IBAction)Operator_pressed:(UIButton*)sender {
//  Get the operator from user, set it as a string and save in newSign
    NSString *newSign = sender.titleLabel.text;
    
    //Call the calculate function and send parameter newSign and
    //store in the variable result.
     double result = [self.myModel calculate:newSign];
    
  // change the format to string and display it
    self.myDisplay.text = [NSString stringWithFormat:@"%0.2f",result];
    
    // change the bool to yes so we can clear previous result and enter a new one
    self.isNewNumber = YES;
    

   
}
// Function used to push values to array and clear the display
- (IBAction)Enter_pressed:(UIButton*)sender {
    // Check if the display is empty or not by checking its length
    if(self.myDisplay.text.length > 0){
        // if not empty get the value and store it as  string
        NSString *valueOnDisplay = self.myDisplay.text;
        
        // Call pushItem function from calBrain and
        //push the value to the array
        [self.myModel pushItem:(valueOnDisplay.doubleValue)];
        
        // clear the display
        self.myDisplay.text = @"";
    }

    
  
}
- (IBAction)Clear_pressed:(UIButton *)sender {
    // check if the display is empty or not
    if(self.myDisplay.text.length >0){
    // clear what in the display when clicked
        self.myDisplay.text = @"";
    }
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    // change the boolean value to yes so we can enter a new value
    self.isNewNumber = YES;
}



@end

File: iOS/Swift/Lab1/Calculator/Calculator_MahaB/ViewController.swift
//
//  ViewController.swift
//  Calculator_MahaB
//
//  Created by Maha Basheikh on 2023-10-05.

// This calculator runs as a postfix calculator

import UIKit

class ViewController: UIViewController {
    var isNewNumber = true // boolean to check if the number is a new number
    var mainAnswerDisplay: String? // variable for the main answer
    var expandedAnswerDisplay: String? // variable for the extended answer showing the full operation (not implemented)
    var newDigit: String? // new digit in the number
    var result: Double = 0.0 // result
    //var newNumber: String?

    @IBOutlet weak var expandedAnswerView: UILabel!  // label showing the extended answer for the full operation (not implemented)
    @IBOutlet weak var answerView: UILabel! // label for the main answer
    override func viewDidLoad() {
        super.viewDidLoad()
        isNewNumber = true // setting the isNewNumber to true
        
        // Do any additional setup after loading the view.
    }
    
    lazy var brain = CalBrain() // lazy loading the CalBrain
    
    // Creates a button which shows the button value in the main answer label when pressed
    @IBAction func digitPressed(_ sender: UIButton) {
        // if the label is empty
        if answerView.text!.isEmpty {
            isNewNumber = true // set the boolean to the number is new
        }
        else {
            isNewNumber = false // otherwise set the boolean to the number is not new
        }
        
        if isNewNumber { // if the number is new
            answerView.text = sender.currentTitle // save the text from the label in a string
            isNewNumber = false // set the boolean value to the number is not new
        }
        else {
            // if not a new number, assign string value to what's shown on the main answer display
            mainAnswerDisplay = answerView.text
            
            // For each new digit, assign a string value
            newDigit = sender.currentTitle
            
            //unwrapping mainAnswerDisplay optional
            if let possibleMainAnswerDisplay = mainAnswerDisplay, let possibleNewDigit = newDigit {
                let newNumber = possibleMainAnswerDisplay + possibleNewDigit
                
            // display the new value
            answerView.text = newNumber
            }
        }
    }
    
    // When the decimal point button is pressed
    @IBAction func decimalPointPressed(_ sender: UIButton) {
        if answerView.text!.isEmpty { // if the main answer label is not empty
            isNewNumber = true // set the boolean value to the number is new
        }
        else {
            isNewNumber = false // set the boolean value to the number is NOT new
        }
        
        if isNewNumber { // if the number is new
            answerView.text = sender.currentTitle // save the text from the label in a string
            isNewNumber = false // set the boolean value to the number is not new
        }
        else {
            // if not a new number, assign string value to what's shown on the main answer display
            mainAnswerDisplay = answerView.text
            
            // For each new digit, assign a string value
            newDigit = sender.currentTitle
            
            //unwrapping mainAnswerDisplay optional
            if let possibleMainAnswerDisplay = mainAnswerDisplay, let possibleNewDigit = newDigit {
                let newNumber = possibleMainAnswerDisplay + possibleNewDigit
                
            // display the new value
            answerView.text = newNumber
            }
        }
    }
    
    // When an operator button is pressed
    @IBAction func operatorPressed(_ sender: UIButton) {
        var newSign: String // declare a string to store the operator
        newSign = sender.currentTitle! // assign the operator from the main answer label to the string
        
        if let possibleResult = brain.calculate(opr: newSign) { // unwrap the string variable and perform the calculation. Store the result
            
            //let string = String(format: "%.2f", number)
                                    
            answerView.text = String(format:"%.2f", possibleResult) // output the result to the main answer label
            isNewNumber = true // set the boolean value to the number is new
        }
    }
    
    // Deletes just one digit from the answerView display
    @IBAction func backspacePressed(_ sender: UIButton) {
        if !answerView.text!.isEmpty { // if the main answer label is not empty
                        
            mainAnswerDisplay = answerView.text // transfer to a string variable
            
            // Count the number of characters in the string
            //let numberLength = mainAnswerDisplay!.count
            
            // Remove the last character in the string
            mainAnswerDisplay!.removeLast(1)
                        
            // Replace the string in the label
            answerView.text = mainAnswerDisplay!
        }
    }
    
    // Changing the sign of the number entered
    @IBAction func changeSignPressed(_ sender: UIButton) {
        if !answerView.text!.isEmpty { // if the main answer label is not empty
            mainAnswerDisplay = answerView.text // extract the text from it and store it in a string
            
            let resultSignChanged = Double (mainAnswerDisplay!) // Take the string, force unwrap it, cast it as a double and store it in a new variable
            
            answerView.text = String(format:"%.2f", resultSignChanged! * -1) // multiply that variable and output it to the main answer label
        }
    }
    
    // Button to clear the display completely
    @IBAction func deletePressed(_ sender: UIButton) {
        if !answerView.text!.isEmpty { // if the main answer label is not empty
            answerView.text = "" // wipe it clean
        }
        
    }
    
    // When the equal button is pressed
    @IBAction func equalPressed(_ sender: UIButton) {
        if !answerView.text!.isEmpty { // if the main answer display is not empty
            
            mainAnswerDisplay = answerView.text // extract its text and store it in a string
            
            if let possible_mainAnswerDisplay = mainAnswerDisplay { // unwrap that string and cast it as a double
                let resultDouble = Double (possible_mainAnswerDisplay)
                
                brain.pushItem(number: resultDouble!) // push that double onto the array
                
                answerView.text = "" // clear the main answer display to for the next entry
            }
        }
    }
    
    
}

